File: py_version/splash.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# splash.py
#
# 2014-01-30 -- created
# 2015-11-11 -- last updated
#
# ~~~~~~~~~
# citation:
# ~~~~~~~~~
# T. W. Davis, I. C. Prentice, B. D. Stocker, R. J. Whitley, H. Wang, B. J.
# Evans, A. V. Gallego-Sala, M. T. Sykes, and W. Cramer, Simple process-led
# algorithms for simulating habitats (SPLASH): Modelling radiation evapo-
# transpiration and plant-available moisture, Geoscientific Model Development,
# 2015 (in progress)

###############################################################################
## IMPORT MODULES:
###############################################################################
import numpy

from const import kCw, kWm
from evap import EVAP


###############################################################################
## CLASSES
###############################################################################
class SPLASH:
    """
    Name:     SPLASH
    Features: This class updates daily quantities of radiation,
              evapotranspiration, soil moisture and runoff based on SPLASH.
    """
    # \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
    # Class Initialization
    # ////////////////////////////////////////////////////////////////////////
    def __init__(self, lat, elv):
        """
        Name:     SPLASH.__init__
        Input:    - float, latitude, degrees (lat)
                  - float, elevation, meters (elv)
        """
        # Error handle and assign required public variables:
        self.elv = elv
        #
        if lat > 90.0 or lat < -90.0:
            print "Latitude outside range of validity (-90 to 90)!"
            exit(1)
        else:
            self.lat = lat
        #
        # Initialize daily status variables:
        self.ho = 0.      # daily solar irradiation, J/m2
        self.hn = 0.      # daily net radiation, J/m2
        self.ppfd = 0.    # daily PPFD, mol/m2
        self.cond = 0.    # daily condensation water, mm
        self.wn = 0.      # daily soil moisture, mm
        self.precip = 0.  # daily precipitation, mm
        self.ro = 0.      # daily runoff, mm
        self.eet = 0.     # daily equilibrium ET, mm
        self.pet = 0.     # daily potential ET, mm
        self.aet = 0.     # daily actual ET, mm

    # \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
    # Class Function Definitions
    # ////////////////////////////////////////////////////////////////////////
    def spin_up(self, d):
        """
        Name:     SPLASH.spin
        Input:    - DATA class, (d)
                  - bool, (to_write)
        Output:   None.
        Features: Spins up the daily soil moisture.
        Depends:  quick_run
        """
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 1. Create a soil moisture array:
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        if isinstance(d.num_lines, list):
            n = d.num_lines[0]
            if (numpy.array(d.num_lines) == n).all():
                wn_vec = numpy.zeros((n,))
            else:
                print "Invalid number of lines read from DATA class!"
        else:
            n = d.num_lines
            wn_vec = numpy.zeros((n,))
        print "Created soil moisture array of length", len(wn_vec)
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 2. Run one year:
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        for i in xrange(n):
            # Get preceding soil moisture status:
            if i == 0:
                wn = wn_vec[-1]
            else:
                wn = wn_vec[i-1]
            #
            # Calculate soil moisture and runoff:
            sm, ro = self.quick_run(n=i+1,
                                    y=d.year,
                                    wn=wn,
                                    sf=d.sf_vec[i],
                                    tc=d.tair_vec[i],
                                    pn=d.pn_vec[i])
            wn_vec[i] = sm
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 3. Calculate change in starting soil moisture:
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        start_sm = wn_vec[0]
        end_sm, ro = self.quick_run(n=1,
                                    y=d.year,
                                    wn=wn_vec[-1],
                                    sf=d.sf_vec[0],
                                    tc=d.tair_vec[0],
                                    pn=d.pn_vec[0])
        diff_sm = numpy.abs(end_sm - start_sm)
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 4. Equilibrate:
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        spin_count = 1
        while diff_sm > 1.0:
            for i in xrange(n):
                # Get preceding soil moisture status:
                if i == 0:
                    wn = wn_vec[-1]
                else:
                    wn = wn_vec[i-1]
                #
                # Calculate soil moisture and runoff:
                sm, ro = self.quick_run(n=i+1,
                                        y=d.year,
                                        wn=wn,
                                        sf=d.sf_vec[i],
                                        tc=d.tair_vec[i],
                                        pn=d.pn_vec[i])
                wn_vec[i] = sm
            #
            start_sm = wn_vec[0]
            end_sm, ro = self.quick_run(n=1,
                                        y=d.year,
                                        wn=wn_vec[-1],
                                        sf=d.sf_vec[0],
                                        tc=d.tair_vec[0],
                                        pn=d.pn_vec[0])
            diff_sm = numpy.abs(end_sm - start_sm)
            spin_count += 1
        #
        print "Spun", spin_count, "years"
        self.wn_vec = wn_vec
        self.wn = wn_vec[-1]

    def quick_run(self, n, y, wn, sf, tc, pn):
        """
        Name:     SPLASH.quick_run
        Inputs:   - int, day of year (n)
                  - int, year (y)
                  - float, daily soil water content, mm (wn)
                  - float, daily fraction of bright sunshine (sf)
                  - float, daily air temperature, deg C (tc)
                  - float, daily precipitation, mm (pn)
        Output:   - float, soil moisture, mm (sm)
                  - float, runoff, mm (ro)
        Features: Returns daily soil moisture and runoff.
        Depends:  - kCw
                  - kWm
                  - EVAP
        """
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 1. Calculate evaporative supply rate, mm/h
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        sw = kCw*(wn/kWm)
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 2. Calculate radiation and evaporation quantities
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        my_evap = EVAP(self.lat, n, self.elv, y, sf, tc, sw)
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 3. Calculate today's soil moisture, mm
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        sm = wn + pn + my_evap.cond - my_evap.aet_d
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 4. Calculate runoff, mm
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        if sm > kWm:
            # Bucket is too full
            #   allocate excess water to runoff
            #   set soil moisture to capacity (i.e., kWm)
            ro = sm - kWm
            sm = kWm
        elif sm < 0:
            # Bucket is too empty
            #   set soil moisture and runoff to zero
            sm = 0
            ro = 0
        else:
            ro = 0
        #
        return(sm, ro)

    def run_one_day(self, n, y, wn, sf, tc, pn):
        """
        Name:     SPLASH.run_one_day
        Inputs:   - int, day of year (n)
                  - int, year (y)
                  - float, soil water content, mm (wn)
                  - float, fraction of bright sunshine (sf)
                  - float, air temperature, deg C (tc)
                  - float, precipitation, mm (pn)
        Outputs:  None
        Features: Runs SPLASH model for one day.
        Depends:  - kCw
                  - kWm
                  - EVAP
        """
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 0. Set meteorological variables:
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        self.precip = pn    # daily precipitation, mm
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 1. Calculate evaporative supply rate (sw), mm/h
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        sw = kCw*(wn/kWm)
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 2. Calculate radiation and evaporation quantities
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        my_evap = EVAP(self.lat, n, self.elv, y, sf, tc, sw)
        self.ho = my_evap.ra_d      # daily solar irradiation, J/m2
        self.hn = my_evap.rn_d      # daily net radiation, J/m2
        self.ppfd = my_evap.ppfd_d  # daily PPFD, mol/m2
        self.cond = my_evap.cond    # daily condensation water, mm
        self.eet = my_evap.eet_d    # daily equilibrium ET, mm
        self.pet = my_evap.pet_d    # daily potential ET, mm
        self.aet = my_evap.aet_d    # daily actual ET, mm
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 3. Calculate today's soil moisture (sm), mm
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        sm = wn + pn + my_evap.cond - my_evap.aet_d
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 4. Calculate runoff (ro), mm
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        if sm > kWm:
            # Bucket is too full
            #   allocate excess water to runoff
            #   set soil moisture to capacity (i.e., kWm)
            ro = sm - kWm
            sm = kWm
        elif sm < 0:
            # Bucket is too empty
            #   reduce actual ET by discrepancy amount
            #   set soil moisture and runoff to zero
            self.aet += sm
            sm = 0
            ro = 0
        else:
            ro = 0
        #
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        # 5. Update soil moisture & runoff
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        self.wn = sm  # daily soil moisture, mm
        self.ro = ro  # daily runoff, mm

    def print_vals(self):
        """
        Name:     SPLASH.print_vals
        Inputs:   None
        Outputs:  None
        Features: Prints all daily values
        """
        print "Daily values:"
        print "  Ho: %0.6f  MJ/m^2" % ((1.0e-6)*self.ho)
        print "  HN: %0.6f MJ/m^2" % ((1.0e-6)*self.hn)
        print "  PAR: %0.6f mol/m^2" % (self.ppfd)
        print "  Cn: %0.6f mm" % (self.cond)
        print "  EET: %0.6f mm" % (self.eet)
        print "  PET: %0.6f mm" % (self.pet)
        print "  AET: %0.6f mm" % (self.aet)
        print "  Wn: %0.6f mm" % (self.wn)
        print "  RO: %0.6f mm" % (self.ro)

File: r_version/splash.R
# R version 3.2.3 (2015-12-10) -- "Wooden Christmas-Tree"
#
# splash.R
#
# last updated: 2016-01-22
#
# ~~~~~~~~~
# citation:
# ~~~~~~~~~
# T. W. Davis, I. C. Prentice, B. D. Stocker, R. J. Whitley, H. Wang, B. J.
# Evans, A. V. Gallego-Sala, M. T. Sykes, and W. Cramer, Simple process-led
# algorithms for simulating habitats (SPLASH): Robust indices of radiation
# evapo-transpiration and plant-available moisture, Geoscientific Model
# Development, 2016 (in progress)
#
# ~~~~~~~~~~~~
# description:
# ~~~~~~~~~~~~
# This script contains functions for running SPLASH for point-based data, i.e.:
#   spin_up(list mdat, list dtot)
#   quick_run(double lat, double elv, double n, double y, double wn, double sf,
#             double tc, double pn)
#   run_one_day(double lat, double elv, double n, double y, double wn,
#               double sf, double tc, double pn)
#
# ~~~~~~~~~~
# changelog:
# ~~~~~~~~~~
# - added fix to daily soil moisture when n and ny are 365 [15.01.27]
#
#### IMPORT SOURCES ##########################################################
source("const.R")
source("evap.R")


#### DEFINE FUNCTIONS ########################################################
# ************************************************************************
# Name:     spin_up
# Inputs:   - list, meteorological data (mdat)
#           - list, daily totals (dtot)
# Returns:  list, meteorological data
# Features: Updates the soil moisture in daily totals until equilibrium
# Depends:  quick_run
# ************************************************************************
spin_up <- function(mdat, dtot) {
    # Run one year:
    for (i in seq(from=1, to=mdat$num_lines, by=1)) {
        if (i == 1) {
            wn <- dtot$wn[mdat$num_lines]
        } else {
            wn <- dtot$wn[i - 1]
        }
        my_vals <- quick_run(mdat$lat_deg, mdat$elv_m, i, mdat$year, wn,
                             mdat$sf[i], mdat$tair[i], mdat$pn[i])
        dtot$wn[i] <- my_vals$sm
    }

    # Calculate the change:
    start_sm <- dtot$wn[1]
    end_vals <- quick_run(mdat$lat_deg, mdat$elv_m, 1, mdat$year,
                          dtot$wn[mdat$num_lines], mdat$sf[1], mdat$tair[1],
                          mdat$pn[1])
    diff_sm <- abs(end_vals$sm - start_sm)

    # Equilibrate:
    spin_count <- 1
    while (diff_sm > 1.0) {
        for (i in seq(from=1, to=mdat$num_lines, by=1)) {
            if (i == 1) {
                wn <- dtot$wn[mdat$num_lines]
            } else {
                wn <- dtot$wn[i - 1]
            }
            my_vals <- quick_run(mdat$lat_deg, mdat$elv_m, i, mdat$year, wn,
                                 mdat$sf[i], mdat$tair[i], mdat$pn[i])
            dtot$wn[i] <- my_vals$sm
        }

        # Calculate the change:
        start_sm <- dtot$wn[1]
        end_vals <- quick_run(mdat$lat_deg, mdat$elv_m, 1, mdat$year,
                              dtot$wn[mdat$num_lines], mdat$sf[1],
                              mdat$tair[1], mdat$pn[1])
        diff_sm <- abs(end_vals$sm - start_sm)
        spin_count <- spin_count + 1
    }
    cat(paste("Spun", spin_count, "years\n"))
    return(dtot)
}


# ************************************************************************
# Name:     quick_run
# Inputs:   - double, latitude, deg (lat)
#           - double, elevation, m (elv)
#           - double, day of year (n)
#           - double, year (y)
#           - double, daily soil moisture content, mm (wn)
#           - double, daily fraction of bright sunshine (sf)
#           - double, daily air temperature, deg C (tc)
#           - double, daily precipitation, mm (pn)
# Returns:  list
#             $sm - soil moisture, mm
#             $ro - runoff, mm
# Features: Returns daily soil moisture and runoff
# Depends:  evap
# ************************************************************************
quick_run <- function(lat, elv, n, y, wn, sf, tc, pn) {
    # Calculate evaporative supply (mm/hr)
    sw <- kCw*wn/kWm

    # Compute daily radiation and evaporations values:
    ET <- calc_daily_evap(lat, n, elv, y, sf, tc, sw)

    # Update daily soil moisture:
    sm <- wn + pn + ET$cond_mm - ET$aet_mm

    if (sm > kWm) {
        # Bucket is full:
        # - set soil moisture to capacity
        # - add remaining water to runoff
        ro <- sm - kWm
        sm <- kWm
    } else if (sm < 0) {
        # Bucket is empty:
        # - set runoff and soil moisture equal to zero
        ro <- 0
        sm <- 0
    } else {
        ro <- 0
    }

    rval <- list()
    rval$sm <- sm
    rval$ro <- ro
    return(rval)
}


# ************************************************************************
# Name:     run_one_day
# Inputs:   - double, latitude, deg (lat)
#           - double, elevation, m (elv)
#           - double, day of year (n)
#           - double, year (y)
#           - double, daily soil moisture content, mm (wn)
#           - double, daily fraction of bright sunshine (sf)
#           - double, daily air temperature, deg C (tc)
#           - double, daily precipitation, mm (pn)
# Returns:  list
#             $ho - daily solar irradiation, J/m2
#             $hn - daily net radiation, J/m2
#             $ppfd - daily PPFD, mol/m2
#             $cond - daily condensation water, mm
#             $eet - daily equilibrium ET, mm
#             $pet - daily potential ET, mm
#             $aet - daily actual ET, mm
#             $wn - daily soil moisture, mm
#             $ro - daily runoff, mm
# Features: Runs SPLASH at a single location for one day.
# Depends:  evap
# ************************************************************************
run_one_day <- function(lat, elv, n, y, wn, sf, tc, pn) {
    # Return values
    rvals <- list()

    # Calculate evaporative supply (mm/hr)
    sw <- kCw*wn/kWm

    # Compute daily radiation and evaporations values:
    ET <- calc_daily_evap(lat, n, elv, y, sf, tc, sw)
    rvals$ho <- ET$ra_j.m2
    rvals$hn <- ET$rn_j.m2
    rvals$ppfd <- ET$ppfd_mol.m2
    rvals$cond <- ET$cond_mm
    rvals$eet <- ET$eet_mm
    rvals$pet <- ET$pet_mm
    rvals$aet <- ET$aet_mm

    # Update daily soil moisture:
    sm <- wn + pn + ET$cond_mm - ET$aet_mm

    if (sm > kWm) {
        # Bucket is full:
        # - set soil moisture to capacity
        # - add remaining water to runoff
        ro <- sm - kWm
        sm <- kWm
    } else if (sm < 0) {
        # Bucket is empty:
        # - reduce actual ET by discrepancy amount
        # - set runoff and soil moisture equal to zero
        rvals$aet <- rvals$aet + sm
        ro <- 0
        sm <- 0
    } else {
        ro <- 0
    }

    rvals$wn <- sm
    rvals$ro <- ro
    return(rvals)
}
